dune/grid/polyhedralgrid/grid.hh

```diff
@@ -0,0 +1,1207 @@
+      const int index = seed.index();
+      switch (codim)
+      {
+        case 0:
+          return cellVertices_[ index ].size();
+        case 1:
+          return 0;//grid_.cell_facepos[ index+1 ] - grid_.cell_facepos[ index ];
+        case dim:
+          return 1;
+      }
+      return 0;
+    }
+
+    template <class EntitySeed>
+    GlobalCoordinate
+    corner( const EntitySeed& seed, const int i ) const
+    {
+      const int codim = EntitySeed :: codimension;
+      switch (codim)
+      {
+        case 0:
+          {
+            const int coordIndex = GlobalCoordinate :: dimension * cellVertices_[ seed.index() ][ i ];
+            return copyToGlobalCoordinate( grid_.node_coordinates + coordIndex );
+          }
+        case 1:
+        {
+          const int faceVertex = grid_.face_nodes[grid_.face_nodepos[seed.index()] + i];
+          return copyToGlobalCoordinate( grid_.node_coordinates + GlobalCoordinate :: dimension * faceVertex );
+        }
+        case dim:
+          {
+            const int coordIndex = GlobalCoordinate :: dimension * seed.index();
+            return copyToGlobalCoordinate( grid_.node_coordinates + coordIndex );
+          }
+      }
+      return GlobalCoordinate( 0 );
+    }
+
+    template <class EntitySeed>
+    int subEntities( const EntitySeed& seed, const int codim ) const
+    {
+      const int index = seed.index();
+      switch (codim)
+      {
+        case 0:
+          return 1;
+        case 1:
+          return grid_.cell_facepos[ index+1 ] - grid_.cell_facepos[ index ];
+        case dim:
+          return cellVertices_[ index ].size();
+      }
+      return 0;
+    }
+
+    template <int codim>
+    typename Codim<codim>::EntitySeed
+    subEntitySeed( const typename Codim<0>::EntitySeed& elemSeed, const int i ) const
+    {
+      assert( i>= 0 && i<subEntities( elemSeed, codim ) );
+      typedef typename Codim<codim>::EntitySeed  EntitySeed;
+      if( codim == 0 )
+        return EntitySeed( elemSeed.index() );
+      else if ( codim == 1 )
+      {
+        return EntitySeed( grid_.cell_faces[ grid_.cell_facepos[ elemSeed.index() ] + i ] );
+      }
+      else if ( codim == dim )
+      {
+        return EntitySeed( cellVertices_[ elemSeed.index() ][ i ] );
+      }
+      else
+      {
+        DUNE_THROW(InvalidStateException,"codimension not availalbe");
+      }
+    }
+
+    const std::vector< GeometryType > &geomTypes ( int codim ) const
+    {
+      static std::vector< GeometryType > emptyDummy;
+      if (0 <= codim && codim < geomTypes_.size())
+        return geomTypes_[codim];
+
+      return emptyDummy;
+    }
+
+    int indexInInside( const typename Codim<0>::EntitySeed& seed, const int i ) const
+    {
+      if( ! grid_.cell_facetag )
+        return i;
+      else
+      {
+        // assert( i>= 0 && i<subEntities( EntitySeed( seed.index() ) ) );
+        return grid_.cell_facetag[ grid_.cell_facepos[ seed.index() ] + i ] ;
```

Comment:
The way I read this, for face `i` of a cell, you return just `i` unless it has a logical cartesian or cornerpoint structure, in which case you return the canonical (0-5) index for the face. So for a cornerpoint grid, at a fault (say in the J+ direction) where the cell has 2 neighbours (and therefore two adjacent faces), a total of 7 faces, the values returned would be:
```
i   indexInInside(i)
0     0
1     1
2     2
3     3
4     3
5     4
6     5
```
If I understand the dune grid interface correctly, the indices returned from indexInInside() should match up with the codim 1 entities of the grid. This means that for the logical-cartesian case (detected by having grid_.cell_facetag being non-null) the code here assumes that the canonical logical cartesian faces are exactly the codim 1 entities of the grid. Whereas in the general case, the codim 1 entities are assumed to be identical to the faces of the UnstructuredGrid. However, this is not reflected in the `subEntities()` and `subEntitySeed()` methods, where the general case seems to be assumed.

The `CpGrid` class handles this by simply saying there are no codim 1 entities. It still returns something (the logical cartesian index) from `Intersection::indexInInside()` though, which may be said to break the interface since there are no subentities of codim 1 to index into.